packages/workbox-cache-expiration/CacheExpirationPlugin.mjs

```diff
@@ -0,0 +1,211 @@
+    await cacheExpiration.expireEntries();
+  }
+
+  /**
+   * Checks whether a `Response` is "fresh", based on the `Response's`
+   * `Date` header and the `maxAgeSeconds` parameter passed into the
+   * constructor.
+   *
+   * The general approach is to default to fresh unless proven otherwise.
+   *
+   * If `maxAgeSeconds` or the `Date` header is not set then it will
+   * default to returning `true`, i.e. the response is still fresh and should
+   * be used.
+   *
+   * @param {Object} input
+   * @param {string} input.cacheName
+   * @param {Response} input.cachedResponse The `Response` object that's been
+   *        read from a cache and whose freshness should be checked.
+   *
+   * Defaults to the current time.
+   * @return {boolean} Either `true` if the response is fresh, or
+   * `false` if the `Response` is older than `maxAgeSeconds` and should no
+   * longer be used.
+   */
+  isResponseFresh({cacheName, cachedResponse}) {
+    // Only bother checking for freshness if we have a valid response and if
+    // maxAgeSeconds is set.
+    if (cachedResponse && this._maxAgeSeconds) {
+      if (process.env.NODE_ENV !== 'production') {
+        core.assert.isInstance(cachedResponse, Response, {
+          moduleName: 'workbox-cache-expiration',
+          className: 'CacheExpirationPlugin',
+          funcName: 'isResponseFresh',
+          paramName: 'cachedResponse',
+        });
+      }
+
+      const now = Date.now();
+      const dateHeader = cachedResponse.headers['date'];
+      if (dateHeader) {
+        const parsedDate = new Date(dateHeader);
+        const headerTime = parsedDate.getTime();
+        // If the Date header was invalid for some reason, parsedDate.getTime()
+        // will return NaN. We want to treat that as a fresh response, since we
+        // assume fresh unless proven otherwise.
+        if (isNaN(headerTime)) {
+          return true;
+        }
+
+        // If we have a valid headerTime, then our response is fresh iff the
+        // headerTime plus maxAgeSeconds is greater than the current time.
+        return (headerTime + (this._maxAgeSeconds * 1000)) > now;
+      } else {
+        // TODO (jeffposnick): Change this method interface to be async, and
+        // check for the IDB for the specific URL in order to determine
+        // freshness when Date is not available.
+
+        // If there is no Date header (i.e. if this is a cross-origin response),
+        // then we don't know for sure whether the response is fresh or not.
+        // One thing we can do is trigger cache expiration, which will clean up
+        // any old responses based on IDB timestamps, and ensure that when a
+        // cache-first handler is used, stale responses will eventually be
+        // replaced (though not until the *next* request is made).
+        // See https://github.com/GoogleChrome/workbox/issues/691
+
+        const cacheExpiration = this._getCacheExpiration(cacheName);
```

Comment:
Even though it's not going to block the `respondWith()`, there's extra asynchronous code that's going to run as the result of this, and if there is a burst of requests, those asynchronous tasks will all be scheduled in short succession.

This was less of a concern in v2, since we used a mutex to ensure that only one expiration job happened at any given time; see https://github.com/GoogleChrome/workbox/issues/922

Reply:
will add it in during the next PR.

This code as a whole is just full of edge cases and unpredictability so the mutex is needed.